src/main/java/frc/trigon/robot/poseestimation/robotposesources/RobotPoseSource.java

```diff
@@ -92,13 +135,40 @@ private boolean isNewTimestamp() {
         return true;
     }
 
-    private Pose3d doubleArrayToPose3d(double[] doubleArray) {
-        if (doubleArray == null || doubleArray.length != 6)
-            return null;
+    private void logVisibleTags() {
+        if (inputs.hasResult) {
+            Logger.recordOutput("VisibleTags/" + this.getName(), new Pose2d[0]);
+            return;
+        }
 
-        return new Pose3d(
-                new Translation3d(doubleArray[0], doubleArray[1], doubleArray[2]),
-                new Rotation3d(doubleArray[3], doubleArray[4], doubleArray[5])
-        );
+        final Pose2d[] visibleTagPoses = new Pose2d[inputs.visibleTagIDs.length];
+        for (int i = 0; i < visibleTagPoses.length; i++)
+            visibleTagPoses[i] = FieldConstants.TAG_ID_TO_POSE.get(i).toPose2d();
+        Logger.recordOutput("VisibleTags/" + this.getName(), visibleTagPoses);
+    }
+
+    private Matrix<N3, N1> calculateSolvePNPStdDevs() {
+        final int numberOfVisibleTags = inputs.visibleTagIDs.length;
+        final double translationStd = translationsStdExponent * inputs.averageDistanceFromAllTags * inputs.averageDistanceFromAllTags / numberOfVisibleTags;
+        final double thetaStd = thetaStdExponent * Math.pow(inputs.averageDistanceFromAllTags, 2) / numberOfVisibleTags;
+
+        return VecBuilder.fill(translationStd, translationStd, thetaStd);
+    }
+
+    /**
+     * Calculates the standard deviations of the pose.
```

Comment:
It calculates the standard deviations of the pose, for when the pose is calculated assuming a robot heading